lib/elixir/lib/calendar.ex

```diff
@@ -99,6 +123,31 @@ defmodule Calendar do
   """
   @callback datetime_to_string(year, month, day, hour, minute, second, microsecond,
                                time_zone, zone_abbr, utc_offset, std_offset) :: String.t
+
+  @doc """
+  Converts the given datetime (with time zone)
+  into the internal Calendar.rata_die format.
+  """
+  @callback datetime_to_rata_die(year, month, day, hour, minute, second, microsecond,
```

Comment:
> I am worried about making it a responsibility of the calendar because it will most likely end-up that all calendars will have to implement the similar timezone conversion logic.

I think this is not correct, as the timezones with `utc_offset` (relative to UTC) and `std_offset` (used for things like daylight savings time) are something that is only part of the ISO8601 (and derived calendars). And therefore, I would also consider timezone handling to be something that should be part of a calendar module, rather than DateTime.

Most other calendars do not take timezones into account. (Because at the time of their creation, it was impossible to communicate near-instantly with [other places where the sun would be at its highest point at a significantly different time](https://en.wikipedia.org/wiki/Time_zone#History).) This indeed means that if we take 'midnight' to mean 'midnight as seen from the Prime Meridian in Greenwich' that calculations from another calendar, such as the Buddhist calendar, will be relative to that time (that is, relative to UTC), rather than relative to the timezones of Thailand.

So rather, in that case you want to take the time in Thailand, discard the timezone information, and then convert it to the Buddhist calendar.

This does feel weird. I think this problem (timezones across calendars) needs more thought.